sno/apply.py

```diff
@@ -0,0 +1,146 @@
+        # add geometry in
+        r[dataset.geom_column_name] = ogr_to_geom(
+            ogr.CreateGeometryFromJson(json.dumps(d['geometry']))
+        )
+    return r
+
+
+@click.command()
+@click.pass_context
+@click.option(
+    "--commit/--no-commit", "commit", default=True, help="Commit changes",
+)
+@click.option(
+    "--allow-empty",
+    is_flag=True,
+    default=False,
+    help=(
+        "Usually recording a commit that has the exact same tree as its sole "
+        "parent commit is a mistake, and the command prevents you from making "
+        "such a commit. This option bypasses the safety"
+    ),
+)
+@click.argument("patch_file", type=click.File('r', encoding="utf-8"))
+def apply(ctx, *, commit, patch_file, allow_empty, **kwargs):
+    """
+    Applies and commits the given JSON patch (as created by `sno show --json`)
+    """
+    try:
+        patch = json.load(patch_file)
+        json_diff = patch['sno.diff/v1']
+    except (KeyError, json.JSONDecodeError):
+        raise click.FileError("Failed to parse JSON patch file")
+
+    repo = ctx.obj.repo
+    rs = RepositoryStructure(repo)
+    wc = WorkingCopy.open(repo)
+    if not commit and not wc:
+        # TODO: might it be useful to apply without committing just to *check* if the patch applies?
+        raise NotFound("--no-commit requires a working copy", exit_code=NO_WORKING_COPY)
+
+    if wc and wc.is_dirty():
+        raise InvalidOperation(
+            "You have uncommitted changes in your working copy. Commit or discard first"
+        )
+
+    diff = Diff(None)
+    for ds_name, ds_diff_dict in json_diff.items():
+        meta_changes = ds_diff_dict.get('metaChanges', {})
+        if meta_changes:
+            raise NotYetImplemented(
+                "Patches containing schema changes are not yet handled"
+            )
+
+        feature_changes = ds_diff_dict['featureChanges']
+        if not feature_changes:
+            continue
+
+        inserts = []
+        updates = {}
+        deletes = {}
+        dataset = rs.get(ds_name)
+        for change in feature_changes:
+            old = ungeojson_feature(dataset, change.get('-'))
+            new = ungeojson_feature(dataset, change.get('+'))
+            if old and new:
+                # update
+                assert old['fid'] == new['fid']
+                updates[old['fid']] = (old, new)
+            elif old:
+                deletes[old['fid']] = old
+            else:
+                inserts.append(new)
+        diff += Diff(dataset, inserts=inserts, updates=updates, deletes=deletes)
+
+    if commit:
+        if not diff and not allow_empty:
+            raise NotFound("No changes to commit", exit_code=NO_CHANGES)
+        try:
+            metadata = patch['sno.patch/v1']
+        except KeyError:
+            # Not all diffs are patches. If we're given a raw diff, we can't commit it properly
+            # TODO: maybe commit as current user with current timestamp? or maybe don't bother.
```

Comment:
Complicated...

`git apply`: is a fairly low-level patch application command. It actually doesn't create commits at all - only applies patches, and AFAICT it ignores author info altogether.

`git am`: applies patches _from a mailbox_. I've never used it, since I've never tried running git on an actual mailserver... You literally have to pass it a path to a unix mailbox, not a patch file or anything, awkward. I think because `git am` only handles mailboxes it always has authorship info present (the author info isn't present in the patch itself, it comes from the email headers)

Reply:
I've used `git am` before. I think when I was making kawaka and I had to re-import some commits in bulk. You can generate mailbox files with `git format-patch` in bulk so it was a convenient way to bulk move commits.

Reply:
If there's nothing useful to learn from git then lets go with what you have 👍